lib/spack/spack/compilers/msvc.py

```diff
@@ -198,11 +191,34 @@ def __init__(self, *args, **kwargs):
         # paths[2] refers to the fc path and is a generic check
         # for a fortran compiler
         if paths[2]:
+
+            def get_oneapi_root(pth: str):
+                """From within a prefix known to be a oneAPI path
+                determine the oneAPI root path from arbitrary point
+                under root
+
+                Args:
+                    pth: path prefixed within oneAPI root
+                """
+                if not pth or not os.path.basename(pth):
+                    return ""
+                if os.path.basename(pth) == "oneAPI":
+                    return pth
+                return get_oneapi_root(os.path.dirname(pth))
+
             # If this found, it sets all the vars
-            oneapi_root = os.path.join(self.cc, "../../..")
+            oneapi_root = get_oneapi_root(self.fc)
+            if not oneapi_root:
+                raise RuntimeError(f"Non oneAPI Fortran compiler {self.fc} assigned to MSVC")
             oneapi_root_setvars = os.path.join(oneapi_root, "setvars.bat")
+            # some oneAPI versions return a version more precise than their
+            # install paths
```

Comment:
I wanted to explain to anyone reading the code why we're not simply querying the version from the compiler itself rather than parsing its install path (which is what we were doing). I guess if anything it was a note to you and myself during review. Will refactor.

Reply:
I agree with the note, but I think a rewording would help: often comments are intended to explain what you are doing, so I read the comment, then got confused that it appeared to not be doing it, then realized you were probably explaining that there was potential for more checks

Reply:
Agreed the wording is confusing with no context, will refactor, just wanted to be sure we're on the same page re: it's existence.